mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -429,6 +429,22 @@ def delete_run(self, run_id):
             run.lifecycle_stage = LifecycleStage.DELETED
             self._save_to_db(objs=run, session=session)
 
+    def _hard_delete_run(self, run_id):
+        """
+        Permanently delete a run (metadata and metrics, tags, parameters).
+        This is used by the mlflow gc command line and is not intented to be used elsewhere.
+        """
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_id, session=session)
+            session.delete(run)
+
+    def _get_deleted_runs(self):
+        with self.ManagedSessionMaker() as session:
+            return session\
+                .query(SqlRun.run_uuid) \
+                .filter(SqlRun.lifecycle_stage == LifecycleStage.DELETED) \
+                .all()
```

Comment:
The output of this function is a list of `sqlalchemy.util._collections.result` objects, rather than a list of strings (as is the case for the FileStore). As a result, attempting to run `mlflow gc` against a sqlite database causes a failure when an object of type `sqlalchemy.util._collections.result` is fed into `backend_store.get_run()`:

```
Traceback (most recent call last):
  File "/Users/czumar/anaconda2/bin/mlflow", line 11, in <module>
    load_entry_point('mlflow', 'console_scripts', 'mlflow')()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/czumar/mlflow/mlflow/cli.py", line 312, in gc
    run = backend_store.get_run(run_id)
  File "/Users/czumar/mlflow/mlflow/store/tracking/sqlalchemy_store.py", line 416, in get_run
    return run.to_mlflow_entity()
  File "/Users/czumar/anaconda2/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/czumar/mlflow/mlflow/store/db/utils.py", line 80, in make_managed_session
    raise MlflowException(message=e, error_code=INTERNAL_ERROR)
mlflow.exceptions.MlflowException: (sqlite3.InterfaceError) Error binding parameter 0 - probably unsupported type.
[SQL: SELECT runs.run_uuid AS runs_run_uuid, runs.name AS runs_name, runs.source_type AS runs_source_type, runs.source_name AS runs_source_name, runs.entry_point_name AS runs_entry_point_name, runs.user_id AS runs_user_id, runs.status AS runs_status, runs.start_time AS runs_start_time, runs.end_time AS runs_end_time, runs.source_version AS runs_source_version, runs.lifecycle_stage AS runs_lifecycle_stage, runs.artifact_uri AS runs_artifact_uri, runs.experiment_id AS runs_experiment_id
FROM runs
WHERE runs.run_uuid = ?]
[parameters: ((u'3ba20fb8ece546b1a0b734e8924e20b5',),)]
```

Can we make sure that this method returns a list of strings instead? Additionally, can we add end-to-end tests for the `mlflow gc` command against both the file store and the sql store to ensure that the command succeeds against both tracking stores?

Reply:
Done